### PR TITLE
Samx7x port, big-sammy support

### PR DESCRIPTION
**Pull request by perigoso**
This is a WIP on the port for the SAMS70 MCU

usb support deppends on [tinyusb#693](https://github.com/hathach/tinyusb/pull/693)

- [x] cmsis
- [x] family config
- [x] startup
- [x] real world test
- [x] implement io
- [x] implement usb
- [x] ~implement spi~
- [x] ~test sensor~

### CODE REVIEW

**Comment by perigoso** [comment]
TinyUsB driver was merged, so further progress is possible

**Comment by FFY00** [comment]
This should be broken with the new build system changes, let me know if you want me to rebase on main, or if you need any help doing that. The dependency classes changed a little bit, but it should be fairly easy to update.

**Comment by perigoso** [comment]
ok, this is in a minimal working state, ready to being built upon, next is io and usb

**Comment by perigoso** [comment]
blinky

theres only an addressable led on board so this'll have to do for a blinky

**Comment by perigoso** [comment]
I don't think I want to do  anything else here, this is ready.

I get that you dont have time to look at this @FFY00 I'm ok with merging this as is, I think there are no major issues, nothing here is structural to the project so  later cleanup isnt really an issue.

File: src/platform/samx7x/pmc.c [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+	if (bypass) {
+		PMC->CKGR_MOR = (PMC->CKGR_MOR & ~CKGR_MOR_MOSCXTEN_Msk) | CKGR_MOR_KEY_PASSWD | CKGR_MOR_MOSCXTBY_Msk |
+				CKGR_MOR_MOSCSEL_Msk;
+	} else {
+		PMC->CKGR_MOR = (PMC->CKGR_MOR & ~CKGR_MOR_MOSCXTBY_Msk) | CKGR_MOR_KEY_PASSWD | CKGR_MOR_MOSCXTEN_Msk |
+				CKGR_MOR_MOSCXTST(xtal_startup_time);
+
+		/* Wait the Xtal to stabilize */
+		while (!(PMC->PMC_SR & PMC_SR_MOSCXTS_Msk))
+			;
```
Comment (FFY00) [inline src/platform/samx7x/pmc.c:87]: Can you add a `continue` here?

File: src/platform/samx7x/pmc.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	PMC_PCK_UART = 4,
+	PMC_PCK_MCAN = 5,
+	PMC_PCK_TC = 6,
+	PMC_PCK_TC0 = 7,
+};
+
+/*
+ * SPDX-License-Identifier: MIT
+ * SPDX-FileCopyrightText: 2021 Rafael Silva <perigoso@riseup.net>
+ */
```
Comment (FFY00) [inline src/platform/samx7x/pmc.h:51]: This is duplicated.

File: src/platform/samx7x/startup.c [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+extern u32 _etext;
+
+extern u32 _sidata; /* Data source */
+extern u32 _sdata; /* Data destination */
+extern u32 _edata;
+
+extern u32 _sbss; /* BSS destination */
+extern u32 _ebss;
+
+extern u32 _end;
```
Comment (FFY00) [inline src/platform/samx7x/startup.c:27]: `sam.h` defines this, right?

File: src/platform/samx7x/usb.c [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+void usb_init()
+{
+	/* Enable USB peripheral clock */
+	pmc_peripheral_clock_gate(USBHS_CLOCK_ID, 1); // Enable peripheral clock
+
+	/* Wait for the clock to be usable */
+	while (!(USBHS->USBHS_SR & USBHS_SR_CLKUSABLE_Msk))
+		;
```
Comment (FFY00) [inline src/platform/samx7x/usb.c:29]: Add a `continue` here please.

**Review by FFY00 — APPROVED** [review]
The code looks okay, just some minor nits. I don't have the hardware, so can't test, but this looks good to me :+1:

**Comment by FFY00** [comment]
Well, I didn't notice you enable automerge :upside_down_face:

**Comment by FFY00** [comment]
Also, can you update the README?

File: src/platform/samx7x/startup.c (hunk 3 above)
Comment (perigoso) [reply src/platform/samx7x/startup.c:27]: These are provided by the linker

**Comment by perigoso** [comment]
Eh, my bad, i shouldn't have for this

**Comment by perigoso** [comment]
I'll do a couple correction commits for those

File: src/platform/samx7x/pmc.h (hunk 2 above)
Comment (perigoso) [reply src/platform/samx7x/pmc.h:51]: Oh, how did that happen

File: src/platform/samx7x/startup.c (hunk 3 above)
Comment (FFY00) [reply src/platform/samx7x/startup.c:27]: Ah!

**Comment by perigoso** [comment]
done in #99

**Pull request merged**